File: src/Modules/Itms.Modules.Assets/Domain/AssetLifecycle.cs
using System.Collections.Frozen;

namespace Itms.Modules.Assets.Domain;

/// <summary>
/// Which move between asset lifecycle statuses is legal, written once.
/// </summary>
/// <remarks>
/// <para>
/// <b>Keyed on the code, because a status is a configurable row.</b> WP-2.1 made asset
/// statuses reference data an administrator can add to, rename, and retire, so there is no
/// enum to build a table over — but it also gave every status an immutable
/// <see cref="AssetStatusCode"/>, precisely so this table would have something stable to
/// reason about. A rename does not touch these rules; a new custom status is not in them
/// at all, which is the next paragraph.
/// </para>
/// <para>
/// <b>An unknown code has no legal destinations and is not terminal, and both halves
/// matter.</b> Nothing here can name a status somebody invents after this table was
/// written, so a lifecycle transition out of a custom status is refused with a message
/// naming it rather than being waved through into a state machine that does not describe
/// it. But <see cref="IsTerminal"/> answers <see langword="false"/> for the same code,
/// which keeps <em>assignment</em> — a different fact, governed by
/// <see cref="Asset.AssignTo"/> and not by this table — working for an asset in a custom
/// status. Otherwise adding a status would quietly make the equipment in it unissuable.
/// </para>
/// <para>
/// <b>The three terminal statuses are terminal at the human's direction</b> (WP-2.2):
/// <see cref="AssetStatusCode.Retired"/>, <see cref="AssetStatusCode.Lost"/> and
/// <see cref="AssetStatusCode.Disposed"/> have no way out, which is the reading WP-1.3
/// took of SPEC.md's silence for a cancelled ticket. The cost is that an asset retired by
/// mistake has no correction path in this package; putting one in would mean inventing
/// recovery semantics SPEC.md §3 does not define, and an explicit correction workflow can
/// add one deliberately later.
/// </para>
/// <para>
/// <b>This type decides nothing about who may make a move, or what else the move
/// writes.</b> It is a pure lookup over the codes. <see cref="Asset"/> owns the field
/// writes and the invariants that go with them; the endpoint owns the policy.
/// </para>
/// </remarks>
public static class AssetLifecycle
{
    /// <summary>
    /// Every legal destination, keyed by origin. Frozen because it is read on every
    /// transition and never rebuilt.
    /// </summary>
    /// <remarks>
    /// The edges are exactly the ones WP-2.2's five operations need, and no more.
    /// Assignment walks <c>in-stock → deployed</c> and a return walks it back;
    /// <see cref="Asset.SendForRepair"/> walks into <see cref="AssetStatusCode.Repair"/>
    /// from either working state; <see cref="Asset.ReturnToService"/> walks out of it to
    /// whichever state the asset's holder implies; <see cref="Asset.Retire"/> walks into
    /// <see cref="AssetStatusCode.Retired"/> from any of the three. <c>lost</c> and
    /// <c>disposed</c> are present as origins so <see cref="IsTerminal"/> knows them —
    /// nothing in this package moves an asset <em>into</em> either, because WP-2.2 names
    /// no operation that would.
    /// </remarks>
    private static readonly FrozenDictionary<string, FrozenSet<string>> LegalDestinations =
        new Dictionary<string, FrozenSet<string>>(StringComparer.Ordinal)
        {
            [AssetStatusCode.InStock] = Set(AssetStatusCode.Deployed, AssetStatusCode.Repair, AssetStatusCode.Retired),
            [AssetStatusCode.Deployed] = Set(AssetStatusCode.InStock, AssetStatusCode.Repair, AssetStatusCode.Retired),
            [AssetStatusCode.Repair] = Set(AssetStatusCode.Deployed, AssetStatusCode.InStock, AssetStatusCode.Retired),
            [AssetStatusCode.Retired] = FrozenSet<string>.Empty,
            [AssetStatusCode.Lost] = FrozenSet<string>.Empty,
            [AssetStatusCode.Disposed] = FrozenSet<string>.Empty,
        }.ToFrozenDictionary(StringComparer.Ordinal);

    /// <summary>
    /// Whether an asset in <paramref name="from"/> may move to <paramref name="to"/>.
    /// </summary>
    /// <remarks>
    /// A move to the status the asset is already in is <em>not</em> legal. It is not a
    /// no-op: it would raise <c>AssetStatusChanged</c> and write a history entry saying an
    /// asset went from Repair to Repair. The caller asked for something that cannot happen,
    /// and a conflict is the honest answer — the same call WP-1.3 made for a ticket.
    /// </remarks>
    /// <param name="from">The code the asset currently carries.</param>
    /// <param name="to">The code being asked for.</param>
    /// <returns><see langword="true"/> when the move is one this table allows.</returns>
    public static bool CanTransition(string from, string to) =>
        LegalDestinations.TryGetValue(from, out var destinations) && destinations.Contains(to);

    /// <summary>
    /// Whether an asset in <paramref name="code"/> has reached the end of its life.
    /// </summary>
    /// <remarks>
    /// A code this table does not know is <see langword="false"/>, deliberately — see the
    /// remarks on this class. Only the three statuses named as terminal are terminal.
    /// </remarks>
    /// <param name="code">The status code to ask about.</param>
    /// <returns><see langword="true"/> for <c>retired</c>, <c>lost</c>, and <c>disposed</c>.</returns>
    public static bool IsTerminal(string code) =>
        LegalDestinations.TryGetValue(code, out var destinations) && destinations.Count == 0;

    /// <summary>
    /// Every destination legal from <paramref name="from"/>, for a caller that wants to
    /// offer them.
    /// </summary>
    /// <remarks>
    /// WP-2.6 renders the lifecycle actions and its done-criterion says an illegal one is
    /// absent rather than disabled in place. It should read this rather than restate the
    /// table in TypeScript.
    /// </remarks>
    /// <param name="from">The code the asset currently carries.</param>
    /// <returns>The legal destination codes, empty from a terminal or unknown status.</returns>
    public static IReadOnlyCollection<string> DestinationsFrom(string from) =>
        LegalDestinations.TryGetValue(from, out var destinations) ? destinations : FrozenSet<string>.Empty;

    private static FrozenSet<string> Set(params string[] destinations) =>
        destinations.ToFrozenSet(StringComparer.Ordinal);
}

File: src/Itms.Web.Client/src/features/assets/lib/asset-lifecycle.ts
/**
 * Which lifecycle actions an asset offers, derived from what the server said.
 *
 * WP-2.6b's criterion is that an illegal action is *absent* rather than disabled, and that
 * the actions read the server's legal destinations rather than restating
 * `AssetLifecycle`'s table in TypeScript. `AssetLifecycle.DestinationsFrom`'s own doc
 * comment asks for exactly that. So nothing here decides whether a move is legal — it
 * decides only which operation a legal destination corresponds to, and how it is worded.
 *
 * ## The two fields this reads, and why there are two
 *
 * `allowedNextStatusCodes` answers for the lifecycle moves. It cannot answer for
 * assignment, and that is not an oversight in the server: the list is empty both from a
 * terminal status **and** from a custom status an administrator added, and those two cases
 * differ — `Asset.AssignTo` refuses only the terminal three, deliberately, so that adding a
 * status does not quietly make the equipment in it unissuable. `canBeAssigned` is the
 * server's own `!AssetLifecycle.IsTerminal(code)`, computed on every read. Inferring one
 * from the other here would get the custom-status case exactly backwards.
 *
 * ## Neither field is the enforcement
 *
 * The server refuses an illegal move with 409 and a terminal assignment with
 * `assets.asset_not_assignable` whatever a client sends. What this buys is a screen that
 * does not offer a button which always fails.
 */

import type { Asset } from '@/lib/api/types'
import type { AssetLifecycleRoute } from '../api/assets-api'

/** `AssetHistoryEntry.NoteMaxLength`. */
export const noteMaxLength = 1000

/** The six operations SPEC.md §3 names, as a screen offers them. */
export type AssetActionId =
  | 'assign'
  | 'transfer'
  | 'return'
  | 'repair'
  | 'return-to-service'
  | 'retire'

/** One lifecycle action, as a button and the dialog behind it. */
export interface AssetAction {
  readonly id: AssetActionId
  /** The verb on the button. Sentence case, per DESIGN.md §2. */
  readonly label: string
  /** What the dialog says will happen. */
  readonly description: string
  /**
   * Where the write goes: one of the three single-party routes, or `null` for the
   * assignment route, which the three assignment acts share (WP-2.2).
   */
  readonly route: AssetLifecycleRoute | null
  /**
   * What happened, past tense, for the sentence a toast reports it with — "LAP-0042
   * issued", "the asset could not be issued". Written here beside the label so the two
   * cannot describe the same action differently.
   */
  readonly outcome: string
  /** True when the dialog has to collect a person before it can be confirmed. */
  readonly needsHolder: boolean
  /** True for the one action DESIGN.md §4 paints in `danger`. */
  readonly destructive: boolean
}

/**
 * The order the actions are offered in: the move that carries the asset forward from where
 * it stands, then the ways sideways, then the way out. The first is rendered as the
 * screen's primary button.
 */
const offered: readonly AssetActionId[] = [
  'return-to-service',
  'assign',
  'transfer',
  'return',
  'repair',
  'retire',
]

/** The status codes the lifecycle table names, spelled once. */
const inStock = 'in-stock'
const deployed = 'deployed'
const repair = 'repair'
const retired = 'retired'

/**
 * The lifecycle actions this asset offers.
 *
 * @param asset The asset as the server last described it, including its
 * `allowedNextStatusCodes` and `canBeAssigned`.
 * @returns The actions to render, in order. Empty from a terminal status.
 */
export function assetActions(asset: Asset): AssetAction[] {
  const destinations = new Set(asset.allowedNextStatusCodes ?? [])
  const held = asset.assignedToUserId !== null && asset.assignedToUserId !== undefined
  const assignable = asset.canBeAssigned === true

  // Where `POST /returns-to-service` would put the asset: back to whoever holds it, or
  // into stock if nobody does. The server decides this the same way, in
  // `Asset.ReturnToService` — so asking whether *that* destination is legal is asking
  // whether the operation would succeed.
  const backTo = held ? deployed : inStock

  return offered
    .filter((id) => {
      switch (id) {
        case 'assign':
          return assignable && !held
        case 'transfer':
          return assignable && held
        // Taking equipment back clears the holder, and moves the status only when the
        // asset was deployed — which is the one case that needs the destination to be
        // legal.
        case 'return':
          return assignable && held && (asset.assetStatusCode !== deployed || destinations.has(inStock))
        case 'repair':
          return destinations.has(repair)
        case 'return-to-service':
          return destinations.has(backTo)
        case 'retire':
          return destinations.has(retired)
      }
    })
    .map((id) => describe(id, asset, backTo))
}

function describe(id: AssetActionId, asset: Asset, backTo: string): AssetAction {
  switch (id) {
    case 'assign':
      return {
        id,
        label: 'Assign',
        outcome: 'issued',
        description: 'Issue this asset to somebody. Equipment issued out of stock is deployed at the same time.',
        route: null,
        needsHolder: true,
        destructive: false,
      }
    case 'transfer':
      return {
        id,
        label: 'Transfer',
        outcome: 'transferred',
        description:
          'Hand this asset to somebody else. A transfer moves who holds it and nothing else — the equipment does not restart its life.',
        route: null,
        needsHolder: true,
        destructive: false,
      }
    case 'return':
      return {
        id,
        label: 'Return',
        outcome: 'taken back',
        description:
          asset.assetStatusCode === deployed
            ? 'Take this asset back off whoever holds it. It goes into stock.'
            : 'Take this asset back off whoever holds it. Its status does not change — it is not back yet.',
        route: null,
        needsHolder: false,
        destructive: false,
      }
    case 'repair':
      return {
        id,
        label: 'Send for repair',
        outcome: 'sent for repair',
        description:
          'Send this asset away to be fixed. Whoever holds it keeps it on the record, which is what tells a later return where to put it back.',
        route: 'repairs',
        needsHolder: false,
        destructive: false,
      }
    case 'return-to-service':
      // Status-aware wording. "Return to service" is the repair-shop phrase and is what
      // this action almost always is; from anywhere else it is plainly a move to where the
      // destination says, and saying "return to service" there would be misleading.
      return {
        id,
        label:
          asset.assetStatusCode === repair
            ? 'Return to service'
            : backTo === inStock
              ? 'Return to stock'
              : 'Put into service',
        outcome: asset.assetStatusCode === repair ? 'returned to service' : 'brought back',
        description:
          backTo === inStock
            ? 'Bring this asset back into stock. Nobody holds it, so it goes back on the shelf.'
            : 'Bring this asset back into service. It goes back to whoever still holds it.',
        route: 'returns-to-service',
        needsHolder: false,
        destructive: false,
      }
    case 'retire':
      return {
        id,
        label: 'Retire',
        outcome: 'retired',
        description:
          'Take this asset out of service and keep it on the books. Retiring also releases whoever holds it, and it cannot be undone — a retired asset has no way back.',
        route: 'retirements',
        needsHolder: false,
        destructive: true,
      }
  }
}
